### ISSUE TITLE
Make binaries compatible with FreeDOS

### ISSUE THREAD

**Issue by Build-0-Matic** [volkertb/temu-vsb#4]
It would be nice if the Virtual Sound Blaster could be compatible with the JEMM386 memory manager that comes with FreeDOS.
Right now the program simply returns that you need EMM386 or QEMM386 for it to run.

**Comment by volkertb** [comment]
@Build-0-Matic That would be great. But I could really use some help with this. My assembly programming skills are still limited. My first priority has been to make this code build with an open-source assembler, whether that's WASM, JWASM, FASM or NASM. The code isn't truly free as long as it's dependent on closed-source software. Right now, it still only builds properly with TASM.

So far, I've spent quite some time trying to port the code to an assembly dialect that would make it build on those other assemblers, and I've gained experience in doing so, but with my current knowledge and skills I'm not even close to making that work.

Currently, the steps I have in mind are the following:

- Get the code to build with a completely open-source toolchain (assembler and linker)
- Add support for JEMM386
- Add support for additional (more modern) output devices, such as AC'97 ICHx, Intel HD Audio, USB Audio, etc.
- Combine ADLIPT and VSB into a unified emulator that sends FM to an OPLxLPT device in one parallel port, and digital audio to another supported device (a Covox on another parallel port, or one of the aforementioned newer modern output devices)

I believe all of this is doable, but I really need some help from others with this!

Assembly-savvy people I've found on GitHub so far who could perhaps help us out with this:

* @vincentbernat
* @stsp (dosemu2 developer)
* @andrewbird (dosemu2 developer)
* @pdewacht (developer of ADLIPT and other DOS OPLxLPT tools)
* @JKnipperts (forked temu-vsb to create a driver/emulator for the TNDLPT device)
* @Baron-von-Riedesel (a.k.a. Japeth, developer of Jemm386 and JemmEx )
* @jmalak (maintainer of the Open Watcom v2 fork, has been kindly improving JWASM to hopefully make it build temu-vsm)
* @tkchia (does a lot of DOS development stuff, including maintaining a fork of gcc-ia16)
* @PerditionC (FreeDOS kernel developer)
* @vladr (VDMSound developer)

Sorry to mention you guys, but if any of you are willing to help us out with this project, and/or can get us in touch with (more) people who are,  that would be greately appreciated.

Long story short: dust off Andrew Zabolotney's Virtual Sound Blaster TSR from the '90s and make it useful again, by...

* ...making it build with open-source toolchains
* ...making it support more modern sound devices
* ...geting it ready for distribution with FreeDOS

Any help is welcome! Thank you. :innocent:

**Comment by volkertb** [comment]
@Build-0-Matic By the way, in the meantime you could try booting into FreeDOS without having any EMM manager loaded. In other words, with FreeDOS running in actual **real mode** instead of in V86 mode.

The original `vsb.com` executable is actually a combined executable containing two different .com files, one for when QEMM386 is loaded (assembled from `vsb_qemm.asm`) and the other for when no EMM manager is loaded at all (assembled from `vsb_real.asm`).

I understand that this is inconvenient and that many DOS games with which you want to use VSB would work better with access to Expanded Memory. But this way, you can at least test whether VSB still works properly on more modern systems.

I'm curious to hear whether it works for you. Also, if you know of any people with dood DOS x86 assembly skills that would be willing to help us with this project, please get them in touch with me. Thanks. :slightly_smiling_face:

**Comment by Build-0-Matic** [comment]
> 
> 
> @Build-0-Matic That would be great. But I could really use some help with this. My assembly programming skills are still limited.

You're better off than I am as my assembly skills are nonexistent.

I'll try running it with nothing installed so that I get the most memory.

**Comment by volkertb** [comment]
Keep an eye on this issue: https://github.com/open-watcom/open-watcom-v2/issues/615

Once WASM is cable of building https://github.com/volkertb/temu-vsb/blob/master/sbemu/vsb_qemm.asm, that will be an important step towards tackling this, since people willing to assist in the porting effort from QEMM to JEMM will no longer require the proprietary and DOS/Windows-only TASM assembler to build the code.

Then again, anybody savvy enough to port this from a QEMM QPI TSR to a Jemm Loadable Module (JLM) is probably also skilled enough to port the code from TASM dialect to NASM dialect or some other dialect that can be built with an open-source assembler.

Oh, I also opened a feature request to have QPI support added to JEMM386, which would make it compatible with TSRs that currently depend on QEMM386. Not sure how often @Baron-von-Riedesel has time to check these issues and feature requests, though, much less work on them. See https://github.com/Baron-von-Riedesel/Jemm/issues/6

**Comment by stsp** [comment]
What are the goals?
vsb was a very poor emulator, compatible
with a few realmode games, incompatible
with dpmi iirc.
Do you want an sb over pc-speaker?
There are much better ways of doing so.

**Comment by volkertb** [comment]
Thanks for asking!

As far as I know, VSB is the only open source Sound Blaster emulator for DOS that at least somewhat works. (I would gladly be proven wrong on it being the only one, by the way!)

The PC speaker support is obviously not interesting. More interesting is the existing Covox Speech Thing support, the advantage of that being that it's both cheap and would (theoretically) work even in newer post-ISA systems, as long as they still have a DOS-compatible parallel port.

But additionally to the Covox support, I was hoping to use this as a basis on which to build further, for instance by supporting more modern sound output devices.

But you mentioned much better ways of doing so? Could you elaborate?

**Comment by volkertb** [comment]
It might also be useful to emphasize that we are looking for a solution that could be shipped with FreeDOS and would work with an as wide a range of PCs as possible, old and new alike.

**Comment by stsp** [comment]
For old and new pcs I dont have a
solution, no. As a task of replacing vsb,
20 years ago I wrote a pc speaker
driver for linux, and an sb emulator
for dosemu (no dosbox back then).
I wanted to write a civox driver for
linux, but didnt have covox.
I would suggest fullfilling that gap now
and just write a covox driver. Not a
solution for old machines, but good
for new ones.

**Comment by volkertb** [comment]
VSB already supports Covox. But I wanted to make the code independent from proprietary tools such as TASM and TLINK first. And then gradually improve it from there.

Do you think it would be more feasible to abandon the VSB code, instead take the `sb16.c` code from dosemu2 and build a Jemm Loadable Module (JLM) around it that traps the Sound Blaster and Adlib ports? And then send the audio output to a Covox device for now, and then add support for other sound devices later?

And would there be any advantage to doing this in assembly language instead of C? The modern systems that don't have ISA slots anymore are usually fast enough for that not to matter, right?

**Comment by stsp** [comment]
Note that vsb wont work on pre-386
just as well. No solution for old pcs.

**Comment by Build-0-Matic** [comment]
I've just tried it on FreeDOS without anything running from the config.sys file but I still get the QEMM required message. And that is no matter which file I run.

**Comment by volkertb** [comment]
@Build-0-Matic Have you tried `vsb_real.com`? That one should work without QEMM. If not, then I don't understand why there even are two variants of the VSB TSR.

@stsp I agree, it wouldn't be a solution for every possible case.

But older (pre-386) systems usually have ISA slots and you can just install actual Sound Blaster or compatible cards in them. It's mostly a (kind of) solution for the lack of sound support that FreeDOS offers people who try to run it on more modern systems. 

And yes, it would only work with real mode games, at least at first, until Sound Blaster emulation could perhaps be built into a DOS extender like DOS/32A. But that's outside the scope of this project. (It would be a cool follow-up project, though!)

Let's tackle things one at a time. :slightly_smiling_face:

**Comment by stsp** [comment]
Yes, vsb already supports covox, but
adding such driver to linux is not difficult
at all. You'll immediately get dos games
to use it.
As for coding for freedos - my own opinion
is that this is not a 21th century approach.
Which is why I created fdpp - fully 64bit
freedos port that can work on top of linux
kernel, getting use of any sound drivers.
I do not expect everyone to share that
view, but I firmly believe this is the only
possible future of the dos. :)
(currently fdpp is supplied with dosemu2)

**Comment by Build-0-Matic** [comment]
> @Build-0-Matic Have you tried `vsb_real.com`? That one should work without QEMM. If not, then I don't understand why there even are two variants of the VSB TSR.

Tried it too... And still the same result.
It's like if something in FreeDOS itself made VSB think a memory manager was installed.

I run FreeDOS on a netbook off a USB stick. I found a trick for installing it to USB so I don't have to run it in a virtual machine (though it involves installing to VirtualBOX, formatting the USB Stick to FreeDOS Boot using Rufus, opening the VM drive with PeaZip and extracting it's content to the USB.... I know it's not a clean process but it works).

**Comment by volkertb** [comment]
@Build-0-Matic So even `vsb_real.com` complains about QEMM not being loaded? That's really weird, since the README actually stated that that variant was meant for environments without QEMM. :confused: The README mentions an older version of VSB that didn't require QEMU at all. Maybe you can find it online? Otherwise, perhaps you can inquire at forums such as the [DOS forum at VOGONS](https://www.vogons.org/viewforum.php?f=7) for further help on this.

@stsp Yeah, I noticed that project too. And by the way, dosemu2 has been working great so far. It's just that I'm fascinated with the fact that native (16-bit) DOS seems to continue to fully work bare-metal even on the latest PCs, with the exception of sound card support. I think that's really the itch we're trying to scratch here.

Of course, I understand that once newer PCs start to become UEFI-only without legacy BIOS support, running 16-bit DOS bare-metal will finally no longer be possible. And then options like fdpp, virtualization and emulation will remain as the only viable ways to run DOS on modern hardware. And in such cases, emulating Sound Blaster within DOS will indeed become pointless, since it can better be handled by the hypervisor or emulator instead of inside DOS.

**Comment by volkertb** [comment]
@stsp By the way (and this is a bit of topic, I admit), would it be feasible to have fdpp run in a dom0 domain on a type 1 (bare-metal) hypervisor such as Xen? So no Linux kernel underneath, but with privileged direct access to VGA hardware, and integrated Sound Blaster emulation? So UEFI -> Xen -> fdpp (dom0) -> DOS?

It would be a modern and reasonably future-proof solution that would be as close to running on bare-metal as possible, while still providing a solution for the lack of Sound Blaster compatibility on modern systems.

I guess the downside of such a solution is that you wouldn't be able to piggyback on the wide range of supported sound devices that the Linux kernel supports.

**Comment by stsp** [comment]
Ok, if you insist on building it under
freedos, then my best guess is that
this should be done around hdpmi.
Binding it to qemm or emm386, as
the original vsb does, will likely leave
you short of prot-mode games, as
unless dpmi is there, they use vcpi
to load their own ring0 kernel with an
extender around it.
And if you want to do that on C, which
is what you most definitely want if you
intend to port sb16.c of dosemu2, then
you first need to come up with a dpmi
extension to trap io ports.

**Comment by stsp** [comment]
Yes, fdpp is intended to be portable.
The main problem is that v86 mode is
not available on modern cpus, and I
would guess its not available under xen.
In which case you still need dosemu2
to emulate v86.

**Comment by volkertb** [comment]
A DPMI extension to trap I/O ports? Seems way above my abilities. Also, would this work for most of the protected mode games, that run on the DOS/4GW DOS extender? Are they compatible with the HDPMI host, even if such an extension were to be written for it?

As for v86 mode on Xen: if I understand correctly, Xen leverages Intel VT-x and AMD AMD-V hardware virtualization extensions, which allow a guest VM to run in 32-bit mode, including V86 support. Isn't this the same mechanism that dosemu2 also uses through KVM to enable native (non-TCG) execution of 16-bit code on an x86-64 host system? Or doesn't this apply in the privileged dom0 domain?

**Comment by stsp** [comment]
> A DPMI extension to trap I/O ports?

Yes, preferably compatible with those
of emm386 or qemm (which vsb already
uses), to not reinvent the wheel.
Having such extensions, you'll be able
to port sb16.c using djgpp toolchain.

> Are they compatible with the HDPMI host, even if such an extension were to be written for it?

Of course. HDPMI is the one of the most
compatible DPMI impls, and most compatible
means supports most available extenders.

> Isn't this the same mechanism that dosemu2 also uses through KVM to enable native
> (non-TCG) execution of 16-bit code on an x86-64 host system?

Well, yes, dosemu2 _can_ set up v86
via KVM, and also it can emulate v86
or use JIT to re-compile real-mode code
to protected mode.
Neither of these approaches is simple
enough, but of course KVM is much
simpler than jit or emulation. If some
glue between fdpp and KVM is written
that can work with xen, then fine. But
then I suspect that under xen you can't
grab all the needed hardware, anyway
(like timer, interrupt controller etc). And
if you can't grab some hardware, you
need to emulate it.
So I don't believe fdpp can run under
xen directly, but perhaps the needed
wrapper can be much smaller than
dosemu2 on linux. Some wrapper will
be needed.

**Comment by stsp** [comment]
Btw, KVM under linux emulates all the
needed hardware already. If the same is
true under xen, then indeed it might be
possible to run fdpp there.

**Comment by volkertb** [comment]
In HVM mode, Xen uses QEMU to emulate necessary hardware. Additionally, if fdpp were to be running in the dom0 domain, it would also get privileged direct hardware access to the host machine (useful for things like VGA). Shall I open a separate issue in the dosemu2 or fdpp GitHub projects to discuss this further?

As for your suggestion to write a DPMI extension for port trapping, I'll try to do some more research on it. In the meantime, if you can share any links to documentaiton, or perhaps any GitHub projects doing similar things, I would appreciate it. Thanks.

**Comment by stsp** [comment]
I never used xen.
But I very much doubt you can
grab _all_ hardware, like timers
and interrupt controller. Wikipedia
says the paravirtualized kernel
can work in dom0.
Also fdpp requires the C/C++
runtime, which is unlikely to be
available in dom0 (unless you
boot linux there first).
Note that you can grab VGA
also under linux - dosemu can
do that if you disable the linux
modesetting first. So VGA access
is not unique to xen.

> In the meantime, if you can share any links to documentaiton,
> or perhaps any GitHub projects doing similar things, I would appreciate it.

This should be very easy if the DPMI
host runs apps on ring-3 (which I think
is what hdpmi does). In that case, you
only need to make sure the IO bitmap
in TSS is set properly, and you need
to install the GPF handler that will be
called on every port access. Then you
need to decode the access and call
the user's handler. This all can be done
in C with djgpp. The only DPMI extension
you need here, is for the ioperm() call.
We can ask @Baron-von-Riedesel if
what I say is right, and if so - how
difficult is to implement the ioperm()
call in hdpmi (but it can't be too difficult).

**Comment by Baron-von-Riedesel** [comment]
> This should be very easy if the DPMI host runs apps on ring-3 (which I think is what hdpmi does)

Yes

> In that case, you only need to make sure the IO bitmap in TSS is set properly

in standard HDPMI, clients run with CPL & IOPL=3, so no I/O is trapped, and the - one and only - TSS has no IO-bitmap.
It's possible to create a hdpmi exe that runs clients with IOPL=0. In this version, HDPMI "emulates" the IN/OUT opcodes in ring 0. But there's still no IO-bitmap.

> It would be nice if the Virtual Sound Blaster could be compatible with the JEMM386

Jemm386 has an io-bitmap. And there exists a ring-0 extension (JLM) that offers an API to trap port access, just like MS Emm386.

**Comment by stsp** [comment]
But what if under jemm386 the apps
will use vcpi to install their own ring0
dpmi kernel? Which is what I think
they do, unless dpmi is already there.
Or am I missing something?

**Comment by volkertb** [comment]
@Baron-von-Riedesel Instead of having to port VSB to a JLM, Would it be practical to add QEMM (QPI) compatibility to Jemm386? I opened a feature request for that: https://github.com/Baron-von-Riedesel/Jemm/issues/6

**Comment by Baron-von-Riedesel** [comment]
> Or am I missing something?

No, that's quite correct. There's a full context switch whenever the pmode app calls v86-mode, so the io-bitmap of jemm386 is relevant only for "real-mode" apps.

> Would it be practical to add QEMM (QPI) compatibility to Jemm386

I don't know QPI, but I guess it's too much work involved.

> My first priority has been to make this code build with an open-source assembler, whether that's WASM, JWASM, FASM or NASM

jwasm probably would be the best choice. Translating a 16-/32-bit segmented app to FASM would be a hard job.

I looked into the source code and tried to assemble it with jwasm. 

There are a few trivial errors ( ENDP and ENDS without preceding name of the proc/struct), and a few data labels (Desc386) have a trailing ":" ( which makes them code labels in MASM/JWASM style ).
The most problematic thing probably is that a "mixed" group is created, containing both 16-bit and 32-bit code segments. This is NOT allowed in MASM/JWASM, resulting in error "cannot access label through segment registers". It's not hard to circumvent this, but it requires a rather good understanding of assembly language and the segmented memory model.

**Comment by stsp** [comment]
> No, that's quite correct. There's a full context switch whenever the pmode app calls v86-mode,
> so the io-bitmap of jemm386 is relevant only for "real-mode" apps.

Which is why my proposal to implement
it in hdpmi instead. I don't suppose IO
bitmap is difficult to add (isn't it trivial?),
and the rest can be done in a user-supplied
GPF handler, so no other work on hdpmi
side.
Don't you like such plan? :)

**Comment by Baron-von-Riedesel** [comment]
> I don't suppose IO bitmap is difficult to add (isn't it trivial?),

Well, currently the TSS in in conventional memory, so adding an io-bitmap would actually require moving it to extended memory. Additionally, there's no API to access the bitmap from ring 3, so one would have to design and implement one. This makes this job "hard" :))

**Comment by stsp** [comment]
> so adding an io-bitmap would actually require moving it to extended memory.

Just to save 8K of an UMB space?
Oh, this can be optional, so no one
will really care. :)

**Comment by volkertb** [comment]
> I don't know QPI, but I guess it's too much work involved.

Fair enough. I guess it's better to just port the VSB code to a JLM instead. Shall I close that QPI feature request?

Orginally, VSB didn't support any EMM manager for I/O port trapping. It would switch the processor into V86 mode and handle the port trapping stuff itself. That should still be what's happening in the `vsb_real.asm` variant. In a later version, Andrew ported it to work with QEMM (through the QPI API), and the result of that is `vsb_qemm.asm`.

Which one of the two do you think would be easier to use as a basis for porting VSB to a JLM?

> jwasm probably would be the best choice. Translating a 16-/32-bit segmented app to FASM would be a hard job.

Yeah, when I started my efforts with this a few years ago, I foolishly picked FASM as the target dialect first. People in the FASM forum already recommended I try a closer dialect instead. Porting it to NASM also proved too big a hurdle with my limited experience with assembly language.

> I looked into the source code and tried to assemble it with jwasm.
> 
> There are a few trivial errors ( ENDP and ENDS without preceding name of the proc/struct), and a few data labels (Desc386) have a trailing ":" ( which makes them code labels in MASM/JWASM style ).
> The most problematic thing probably is that a "mixed" group is created, containing both 16-bit and 32-bit code segments. This is NOT allowed in MASM/JWASM, resulting in error "cannot access label through segment registers". It's not hard to circumvent this, but it requires a rather good understanding of assembly language and the segmented memory model.

Some help with that would be much appreciated.

It might also be interesting for you to know that @jmalak has been trying to improve WASM (part of his Open Watcom v2 fork) so that it could assemble this code as well. He made quite a bit of progress and already fixed quite a few of the errors. But there's still some ways to go to liberate this code from TASM. Although WASM and JWASM have diverged into two distinct assemblers years ago, it might still be useful to scroll through the messages in that issue thread: https://github.com/open-watcom/open-watcom-v2/issues/615

For instance, he implemented label inference for bare `ENDP` and `ENDS` statements in WASM. Apparently, that's a TASM feature. (Not sure if MASM supports this as well?)

On an unrelated note: I created two feature requests in your Jemm project on GitHub: one for QPI support (which in hindsight, might indeed not be worth the effort, as you mentioned), and the other one for the addition of GEMMIS support, which would allow Windows 3.1 (in 386 Enhanced mode) and Windows 9x to start up from FreeDOS with Jemm386 active. Would that be more feasible to implement? You might be able to reuse the GEMMIS code in DOSBox to implement this.

**Comment by Baron-von-Riedesel** [comment]
> Shall I close that QPI feature request?

yes

> Which one of the two do you think would be easier to use as a basis for porting VSB to a JLM?

The vsb_real.asm thing puts the cpu in v86 mode and then runs as a "micro v86 monitor". That's very interesting - I indeed wonder how stable and compatible that could be - but in no way suitable as a basis for a JLM.

I see that vsp_qemm.asm uses the QPI, but this isn't all. Additionally, it very "hackishly" modifies IRQ0 in the Qemm IDT. I don't understand why this is done and definitely don't recommend doing so in Jemm386. If it's done for speed reasons only, it's nonsense - a 16-bit IRQ0 routine in v86 mode is surely fast enough.

**Comment by volkertb** [comment]
@Baron-von-Riedesel @stsp Okay, although the HDPI extension is definitely an interesting follow-up project that I'd like to explore next, I'd like to continue focusing on VSB first. I could really use some guidance with this, though.

The steps I have currently in mind with this code are the following:

- Get `vsb_qemm.asm` to assemble with an open source assembler (JWASM, WASM, or NASM)
- Port it to a Jemm Loadable Module (JLM)
- Add support for Intel ICHx AC'97 (and compatible) devices as output sinks, using the [ich2player](https://github.com/volkertb/ich2player) code I forked from Jeff Leyda

Would one or both of you be okay with me regularly seeking your help and insight as I try to tackle this? I'm willing to learn as I go. :slightly_smiling_face:

Thanks for considering.

**Comment by stsp** [comment]
First I would suggest to make a list of
games - even only realmode ones - where
vsb actually works.
If my memory serves me well, the emulation
was very incomplete.

**Comment by Baron-von-Riedesel** [comment]
I can surely help with translating the code to masm/jwasm style. However, it would be good if I'm able to run VSB under Qemm386 myself. What is a working configuration? I tried several "recent" machines ( Ryzen 3, i7 4770, Core2Quad ), but Qemm either makes the machine stop when it's loaded or a GPF occurs when VSB.COM is loaded. Does it work in a virtual machine (VBox, Qemu)?

**Comment by stsp** [comment]
Have you seen this project:
https://github.com/charliesome/doslinux
They load linux kernel as a TSR in dos
to use linux drivers.
Might be exactly what you need.

**Comment by volkertb** [comment]
@Baron-von-Riedesel  I have to admit that I was so obsessed in getting the code to assemble properly first, that I hadn't even tried running the originally distributed .COM executable to check whether it would work properly on any modern machines at all. :flushed:

To prevent myself from possibly wasting everyone's time, I'm going to try some virtualization and emulation solutions, and I'll get back to you on this. If for whatever reason the tool only works on older hardware (emulated or not), then it would add an extra hurdle and make the value of resurrecting it and dusting it off even smaller.

I'll also try to enlist help from people on VOGONS and in the FreeDOS mailing lists, for details on how to load the TSR successfully and what games actually work with it. (Although the README included with the source code does contain a list.)

@stsp Although doslinux is indeed a cool project, it seems a bit overkill for this, since my objective here is getting DOS games to work in actual DOS, with sound. Or are you suggesting using doslinux as some kind of backend to which to redirect the sound in a hypothetical Sound Blaster emulator, so I wouldn't have to write DOS drivers for each individual audio device?

Maybe I should start smaller, by first writing some test code to learn how to properly trap I/O port access in an HDPMI extension that would work with both real mode and protected mode games? Without any audio redirection or DMA emulation or any other complex stuff at first? What do you guys think?

**Comment by stsp** [comment]
> Or are you suggesting using doslinux as some kind of backend to which to redirect
> the sound in a hypothetical Sound Blaster emulator, so I wouldn't have to write DOS
> drivers for each individual audio device?

Of course. Especially considering this:
`Add support for additional (more modern) output devices, such as AC'97 ICHx, Intel HD Audio, USB Audio, etc.`
Have you ever written the USB driver for dos? :)

**Comment by volkertb** [comment]
No, but since there is still no proper open source USB driver for DOS, I'm assuming it's far from trivial, right? :sweat_smile:

**Comment by Baron-von-Riedesel** [comment]
> check whether it would work properly on any modern machines at all

I also tried VSB qemm on an old K6-2 machine (1998) and it GPF'ed when loading, so it's probably not a problem of modern machines. I did use Qemm v9 (Qemm97?) Perhaps that's too "new" and VSB requires Qemm v7?

> since my objective here is getting DOS games to work in actual DOS, with sound

Are you aware of this:  [(https://sound-dos.ucoz.ru/load/new_hxdos_extender_2_17/1-1-0-8)](url)

It somewhat works, running dosbox in DOS, although sound is too "delayed" in my opinion.

> how to properly trap I/O port access in an HDPMI extension that would work with both real mode and protected mode games

Such an approach would require to have one io-bitmap for both v86-mode and protected-mode. That's only possible if DPMI is fully integrated into a V86 monitor. It's impossible if the DPMI host is loaded as a VCPI client, as it is the case with HDPMI.

**Comment by raszpl** [comment]
Did DDMA http://intel-vintage-developer.eu5.org/PC-SUPP/PLATFORM/AC97/LEG_PCI.PDF use trapping IO tricks? or was it all hardware based backwards compatibility? afaik all the pci legacy drivers required emm386 and emulation worked in pmode games like Doom/Quake.

**Comment by volkertb** [comment]
How about implementing QPI compatibility (or compatibility with the port trapping API in Microsoft EMM386 for that matter) as a Jemm Loadable Module? Is the JLM architecture flexible enough to extend Jemm like that?

I'm continuing this conversation, since VSB isn't the only TSR that would benefit from this. Another notable one would be SoftMPU. The SoftMPU developers have implemented support for both EMM386 and QEMM, but they don't want to go through the trouble of rewriting it as a JLM in order to add support for Jemm as well.

It would be great if Jemm could be extended modularly like this, without having to introduce any complexity in Jemm itself.

And maybe GEMMIS support could be implemented in the same way?

**Comment by stsp** [comment]
I would advice against both QPI
and EMM386 API for Jemm386.
They both are pretty awkward.
IMHO its better to develop the
proper, DPMI-compatible API
for that (much like emm386's
one, but w/o ring0 descriptors),
and integrate its use into the existing
projects.

**Comment by volkertb** [comment]
> I can surely help with translating the code to masm/jwasm style. However, it would be good if I'm able to run VSB under Qemm386 myself. What is a working configuration? I tried several "recent" machines ( Ryzen 3, i7 4770, Core2Quad ), but Qemm either makes the machine stop when it's loaded or a GPF occurs when VSB.COM is loaded. Does it work in a virtual machine (VBox, Qemu)?

I got it working in an emulated computer of the type `[486] AMI 486 clone` in PCem v17. PCem can also emulate the Covox Speech Thing. It was an MS-DOS 6.22 system with Qemm 7.5. It loaded `VSB` fine and it worked with Wolfenstein 3D (shareware version). The opening of the doors wasn't always played back properly, but most other sound effects sounded fine. I also tried it without QEMM, using its own `micro v86 monitor` mode as you called it, and that worked equally well.

I also tried it on real hardware, a fairly recent Intel motherboard that had a legacy parallel port. I couldn't get QEMM to load on it (it would always crash), but the `micro v86 monitor` mode of VSB worked surprisingly well on this fairly modern bare metal system as well. (The door sounds didn't work at all, though.) I even got it to work with Epic Pinball at one point, although the menu music played back faster that it should, and it wouldn't work anymore after I fiddled with the sound setup settings.

But hey, it works in both emulation and on (modern) bare metal. So it's honestly not that bad an emulation solution to work from. :slightly_smiling_face:

By the way, I've made some progress trying to get the QEMM version of VSB to assemble with WASM (the one that comes with Open Watcom v2), thanks to some crucial help and work by @jmalak. Unfortunately, that build crashed when I tried to load it with QEMM installed, so it's probably better to cherry-pick the specific Sound Blaster emulation code and use that as a basis for developing an emulator that also works with DPMI.

**Comment by volkertb** [comment]
> I would advice against both QPI
> and EMM386 API for Jemm386.
> They both are pretty awkward.
> IMHO its better to develop the
> proper, DPMI-compatible API
> for that (much like emm386's
> one, but w/o ring0 descriptors),
> and integrate its use into the existing
> projects.

I was still kind of hoping to have an "intermediate" solution for real-mode games first, before tackling the apparently much bigger challenge of implementing a DPMI-compatible solution, which apparently has never been done before.



> I also tried VSB qemm on an old K6-2 machine (1998) and it GPF'ed when loading, so it's probably not a problem of modern machines. I did use Qemm v9 (Qemm97?) Perhaps that's too "new" and VSB requires Qemm v7?

I got `vsb_qemm.com` to load with QEMM 97/9.0 as well, at least in a VM in QEMU, since QEMM 97 would always crash when I tried to load it on the bare metal machine that I was also testing with.

> Are you aware of this: [(https://sound-dos.ucoz.ru/load/new_hxdos_extender_2_17/1-1-0-8)](url)

That's for running DOSBox under DOS, right? If I'm going to go that route, I might as well run DOSBox under Linux or Windows. Where's the fun in that? :wink:

> Such an approach would require to have one io-bitmap for both v86-mode and protected-mode. That's only possible if DPMI is fully integrated into a V86 monitor. It's impossible if the DPMI host is loaded as a VCPI client, as it is the case with HDPMI.

So the HDPMI route wouldn't work as a single solution for real-mode games and protected-mode games? You'd need some kind of not-yet-existent alternative for EMM386/QEMM/Jemm that had integrated DPMI? Or am I misunderstanding you?

If not, how likely is it to see DPMI integrated in Jemm any time soon? :innocent:

**Comment by stsp** [comment]
> apparently much bigger challenge of implementing a DPMI-compatible solution, which apparently has never been done before.

Its the same thing, just with a
slightly different API. Since there
are not too many progs that use
either QPI or EMM386 port trapping,
I see no problem at all in implementing
a better API for that, and do a minor
change to all 3 progs that can use it.

> So the HDPMI route wouldn't work as a single solution for real-mode games and protected-mode games?

You don't need to have the same API
for DPMI and real-mode programs.
This is one of the reasons I suggest
an alternative API than that existing.
You need 2 APIs here. Even if very
similar ones, one will be in hdpmi, and
another - in jemm386. The one in
jemm386 should call the _realmode_
handler only.

> If not, how likely is it to see DPMI integrated in Jemm any time soon?

This is why I propose the new API.
In that case, for HDPMI the realmode
API will still work, because the translation
service will install the real-mode callback
there.
So my proposal is to have the API
that works in all cases, and the one
that requires the absolute minimum of
changes in:
- hdpmi
- jemm386
- all 3 users of current EMM386's API
- dosemu2

**Comment by stsp** [comment]
The best way of pursuing such
API is to prototype it in dosemu2
and then present to hdpmi author
to show how small and simple it is.
He may consider implementing it,
or if not - you can do that on your
own.
Waiting forever for a merger of hdpmi/jemm
is a dead end. No one will do such
a tremendous task just to support
some crappy API for 3 progs.

**Comment by volkertb** [comment]
Alright, let's do that. You have a concrete idea how to go about this. Also, you're right. Waiting for others to tackle more complex approaches isn't going to be very fruitful.

So basically, you implement such a prototype API in Dosemu2 and then I try "porting" one of the existing emulation TSRs (adipt, SoftMPU, VSB or TEMU) to it? I'm willing to do that, since it seems like an interesting and educational challenge, but I'll have a lot of questions as I work on it.

Or would you like to suggest a better approach?

**Comment by volkertb** [comment]
ADLIPT might be the best to tackle first, since there's no DMA or IRQ stuff to worry about. It's merely output to two I/O ports that needs to be trapped and diverted to a parallel port device. And I happen to have one of those devices to test with.

**Comment by volkertb** [comment]
Maybe we should "design" the API first? Figure out what stuff we need? For instance, would it be a good idea to implement a generic DMA emulation handler and offer an abstraction for that? Ditto for emulating IRQ triggers, perhaps?

**Comment by stsp** [comment]
Not sure what kind of the "DMA
emulation handler" you need.
The driver programs the ISA DMA
via the port I/O, and that  one you
can trap. Nothing else is needed.
The API is supposed to be really
simple: same as from EMM386,
just split into 2 parts: for rm and
for DPMI.

**Comment by stsp** [comment]
It also seems that the api is needed
to relocate the port ranges. Otherwise
by hooking some range, you will make
it inaccessible even for your own program:
it will trap itself. So the relication API
will help to access the real hw with
another io region.

**Comment by volkertb** [comment]
> Not sure what kind of the "DMA
> emulation handler" you need.
> The driver programs the ISA DMA
> via the port I/O, and that one you
> can trap. Nothing else is needed.
> The API is supposed to be really
> simple: same as from EMM386,
> just split into 2 parts: for rm and
> for DPMI.

Yeah, sorry. This is just my tendency as a high-level programmer to want to create layers of abstractions of everything complex. I was looking at the "pseudo DMA" part, where you hook the timer interrupt and have the handler read and process the RAM as the sound card normally would. But thinking about it again, that problem has already been solved in the existing emulators themselves. You're right that we should stick to the underlying issue here, which is I/O port trapping that works with protected mode software.

> It also seems that the api is needed
> to relocate the port ranges. Otherwise
> by hooking some range, you will make
> it inaccessible even for your own program:
> it will trap itself. So the relication API
> will help to access the real hw with
> another io region.

Sounds like a practical solution. But as a possible alternative to that, wouldn't it be possible to have the emulation handler (the code using the API to intercept those ports) run in a more privileged ring than the application/game that's trying to access the emulated sound card? (On the other hand, your suggestion might be cleaner and safer.)

**Comment by volkertb** [comment]
I believe some modern PC spec by Intel stipulates that all I/O ports below 1000 (not sure if that's 1000 hex or 1000 decimal) are considered legacy ports and are not handed out to PCIe devices that request I/O ports from the chipset. That's also one of the reasons why PCIe sound cards cannot offer DOS-mode hardware compatibility with legacy devices such as parallel ports, etc. 

On modern motherboards, only the integrated Super I/O controller that is connected to the LPC bus maps actual hardware to these legacy I/O ports.

So basically only these sub-1000 I/O ports that are often hard-coded and non-configurable in DOS applications are the ones you'd want to intercept for emulation purposes.

So if you're writing a program that emulates something like a Sound Blaster and redirects the audio to a modern audio device, that audio device will most likely not be listening on any of those legacy ports. So port relocation on the side of the emulator wouldn't be necessary then, right?

**Comment by stsp** [comment]
> So port relocation on the side of the emulator wouldn't be necessary then, right?

Correct.
But there are 2 things to consider:
1. It would be very good to have that
for testing. This way you can add the
SB16 output driver to VSB, that will
only redirect the original IO accesses
to the alias. That way you'll excercise
this new API without writing too much
of the code.
2. You may still want to support some
legacy audio card as an output, like
for example PAS or GUS. And it doesn't
matter if that card can have the different
port range than SB, because the DMA
ports will clash nevertheless.

So while entirely optional in jemm386,
it would be very good to have when
prototyping on dosemu2.

> wouldn't it be possible to have the emulation handler (the code using the API to intercept those ports)
> run in a more privileged ring

Yes this seems possible too.
In jemm386 that will mean IOPL=0
and in dosemu2 that will mean
"temporary remove all trapping
while executing the handler".
So I guess the choice should depend
on what is simpler to implement, and
whether or not the remapping API
can be helpful for something else.

**Comment by stsp** [comment]
> Yes this seems possible too.

This will however prevent the "nesting"
of emulators. Say, you run your VSB
under some covox emulator. :) If it is
running with IOPL=0 then covox emulator
will not do the trick. But with the remapping
API, the covox emulator could feed us
with the alias that he controls. OTOH he
himself will use the alias given by jemm386.
In that chain your VSB will talk to covox
emulator, and covox emulator will talk to
real HW (which is of course also emulated
by dosemu2 or dosbox).
By setting IOPL=0 when running the
handler, you disallow such (not very
realistic) scenario.

**Comment by volkertb** [comment]
That's a very good point. :slightly_smiling_face:

Okay, let's go for the remapping option then.

So what would the next course of action be? Write a draft API with what we think we need, then you try to implement it in the DPMI implementation of Dosemu, and then I try writing or porting an emulator to make use of that API?

And once that works, try to convince developers/maintainers of open source DPMI implementations to adopt the same API? Perhaps we should do some of this stuff in parallel. For instance, we may want to ask input/feedback on the API from the people who maintain HDPMI and other such projects, before putting too much effort into an approach that may not end up gaining any traction later on.

**Comment by volkertb** [comment]
So assuming that I have some kind of DPMI implementation with this API running, and I want to access it from my code, how is this done? I assume I would have to put some 32-bit emulation handling code in resident memory and then call a software interrupt with a pointer to that handler as an argument, and the I/O port I wish to intercept as a second argument? Or would such an API be accessed in a different way than calling an ISR? I'm still learning, so please bear with me. :sweat_smile:

**Comment by stsp** [comment]
> Write a draft API with what we think we need, then you try to implement it in the DPMI

I would suggest to start from the
real-mode part, and forget about
DPMI for a moment. The one is
targeted for jemm386.
On the second step you can add
the translation service to DPMI so
that it installs the real-mode callback
as a handler for the real-mode API.
And only as the last step, consider
the DPMI extension itself.

> So assuming that I have some kind of DPMI implementation with this API running,
> and I want to access it from my code, how is this done?

Not as a first step.
Your VSB is incompatible with
protected mode games even now.
So lets keep it that way for as long
as possible. We need to first put the
existing functionality over the new
(real-mode) API.

**Comment by stsp** [comment]
Not a flawless plan though, as
under dosemu2 you won't be able
to run VSB in protected mode
unless you convert it to DPMI.
Which I suggested to not do as
a first step, so what else to do?
Is there any hope to create a
realmode build of VSB?

**Comment by raszpl** [comment]
> under dosemu2 you won't be able
> to run VSB in protected mode

This thread is getting highly confusing. Why would anyone want to "run VSB in dosemu2"?? dosemu2 already emulates sound blaster https://github.com/dosemu2/dosemu2/blob/devel/src/base/dev/sb16/sb16.c


> until Sound Blaster emulation could perhaps be built into a DOS extender like DOS/32A

This sounds like the most straightforward plan, no drafting of elaborate APIs, just test if its alt all possible by experimenting with dos/32 extender.

Step 1 - make a build environment able to successfully compile https://github.com/abbec/dos32a
Step 2 - hack port trapping into dos/32 to determine viability of the concept on real hardware

**Comment by volkertb** [comment]
> > until Sound Blaster emulation could perhaps be built into a DOS extender like DOS/32A
> 
> This sounds like the most straightforward plan, no drafting of elaborate APIs, just test if its alt all possible by experimenting with dos/32 extender.
> 
> Step 1 - make a build environment able to successfully compile https://github.com/abbec/dos32a Step 2 - hack port trapping into dos/32 to determine viability of the concept on real hardware

Actually, the [Causeway](https://github.com/tkchia/causeway) extender might be a more suitable choice for this. The reason being that DOS/32A runs its application at CPL0 (a.k.a. ring 0, the highest privilege mode, "kernel space", in Linux terms). That makes it impossible for the DOS extender to restrict or intercept any I/O access from the application or game. Causeway, on the other hand, runs its payload at CPL3 ("user space", so to speak), and can therefore trap I/O ports, provided that the DOS extender is running with no external DPMI host already active.

Causeway is also compatible with DOS/4GW, so just like with DOS/32A, it should be possible to use it with most protected mode DOS games as a drop-in replacement. However, I haven't found any information on-line of people who've actually tried that. Plenty of people who've replaced DOS/4GW with DOS/32A, but nothing about trying this with Causeway. It would be interesting to experiment with this first, then.

Anyway, I have a better understanding now about CPL 0 vs CPL 3, and where the IOPL and TSS fit in. And honestly, @stsp's proposal earlier in this thread to implement such emulation in an open source DPMI host instead of in a DOS extender, indeed seems like the best way to go. With such a DPMI host loaded, games with most common DOS extenders would (at least in theory) work with such emulation out of the box, without requiring any game-specific patching.

As a first step, I've been trying to get HDPMI to build in a Linux environment. However, after working on this for quite a few hours, I haven't been successful with that yet.

@Baron-von-Riedesel I see you already added an HDPMI variant to the [HX project](https://github.com/Baron-von-Riedesel/HX) that sets the IOPL to 0 to allow I/O from applications to be intercepted. Thanks for that. However, it seems like the Makefiles of the HX project are really written with DOS and Windows environments in mind. I have a strong preference to develop and cross-compile in Linux. I haven't been successful in building the HDPMI executables on Linux yet. I believe I've gotten somewhat further by making some changes, and even using a case-insensitive folder in Linux, using the new ext4 casefolding feature. Would you have some time in the near future to help me troubleshoot this further? I would appreciate it.

**Comment by stsp** [comment]
You can try Helix netroom's I/O trapping API:
http://www.ctyme.com/intr/int-2c.htm
This one seems to be a very reasonable API.
And perhaps you don't need to even wait for
it to materialize in some open-source DPMI
servers if you can find the helix's manager
over the internet.
I've found it here:
https://winworldpc.com/product/helix-netroom/3x
but its password-protected...

**Comment by stsp** [comment]
So here's the DPMI server with IO trapping
capabilities: https://github.com/dosemu2/dosemu2/files/8016450/nr-nopw.zip
Hope this is what you are looking for.

**Comment by Baron-von-Riedesel** [comment]
> However, it seems like the Makefiles of the HX project are really written with DOS and Windows environments in mind. 

Yes.

>  I haven't been successful in building the HDPMI executables on Linux yet. 

The simplest approach is to run the makefiles and (Win32) binaries (wmake, jwasm, jwlink ) in DosEmu. At least creation of hdpmi should be no problem.

**Comment by raszpl** [comment]
https://www.vogons.org/viewtopic.php?f=62&t=83065

"Here is the current version of the DOS32AWE utility based on the DOS/32A DOS Extender. Now Sound Blaster DOS MIDI synthesizer with Creative's AWEUTIL is also possible for Protected mode DOS games thus allowing you to use MIDI sound banks of your taste while what's best -- you don't need any additional wavetable modules/PCBs/ external MIDI synthesizers, etc. for your ISA AWE Creative Sound Blasters . This utility also assists the VIA's Sound Blaster emulation (VIAFMTSR) that uses VIA chipset south bridges for protected mode games otherwise they hang, fail or sound choppy. If you wonder how to use the DOS32AWE, first ensure the AWEUTIL is loaded as TSR (AWEUTIL /EM:xxx) and then start you game with "DOS32AWE yourgamename" or alternatively replace the DOS4GW .EXE in your games's directory (if exists separately there and is not already bound to the main executable) with the renamed to DOS4GW.EXE DOS32AWE.EXE file.

The DOS32AWE is a DOS32/A DOS Extender modification which is capable of redirecting MIDI sound hardware accesses of DOS Extended applications running in protected mode to the real mode Creative's original AWEUTIL (when loaded as TSR) MIDI emulator."

**Comment by volkertb** [comment]
Thanks for sharing, @raszpl. However, this solution is specific to Sound Blaster AWE cards only. Also, the developer who forked DOS32AWE from DOS/32A hasn't shared the sources of his work, and it's not clear if he's inclined to ever do so. [The license terms of the DOS/32A source code](https://dos32a.narechk.net/content/license.html) technically don't require him to do that. So without those modified sources, we don't have anything to build on further.

It's also worth noting that DOS32AWE depends on the sound card being able to generate NMIs, in order for the emulation to work. It's basically a form of hardware-assisted emulation. It would be nice to have an emulator that doesn't just work in protected mode, but also is flexible in what sound devices it supports, including ones that are not NMI-capable.

**Comment by crazii** [comment]
> > I don't suppose IO bitmap is difficult to add (isn't it trivial?),
> 
> Well, currently the TSS in in conventional memory, so adding an io-bitmap would actually require moving it to extended memory. Additionally, there's no API to access the bitmap from ring 3, so one would have to design and implement one. This makes this job "hard" :))

I think a set of  DPMI "extension" functions could be added, i.e. INT31H, AH=0AH, is that harder or easier?

**Comment by Baron-von-Riedesel** [comment]
The current github hdpmi already has a version with full IO permission bitmap ( and clients run at IOPL 0 ). They have an i-suffix ( hdpmi32i.exe/hdpmi16i.exe ). Also, a "vendor" API has been added ( vendor "HDPMI" ) that is supposed to add port trapping and IRQ virtualization ( not fully specified/implemented yet ).

**Comment by crazii** [comment]
> The current github hdpmi already has a version with full IO permission bitmap ( and clients run at IOPL 0 ). They have an i-suffix ( hdpmi32i.exe/hdpmi16i.exe ). Also, a "vendor" API has been added ( vendor "HDPMI" ) that is supposed to add port trapping and IRQ virtualization ( not fully specified/implemented yet ).

Cool, then it will be much easier to implement port trapping using the existing mechanism.

**Comment by crazii** [comment]
> The current github hdpmi already has a version with full IO permission bitmap ( and clients run at IOPL 0 ). They have an i-suffix ( hdpmi32i.exe/hdpmi16i.exe ). Also, a "vendor" API has been added ( vendor "HDPMI" ) that is supposed to add port trapping and IRQ virtualization ( not fully specified/implemented yet ).

Hi, may I ask another question, are you currently working on the port trapping feature? if not, when is it planned to start/complete?
I have a requirement that to allow non-DPMI client (real mode program or something like that) to install a PM port trap :D

**Comment by Baron-von-Riedesel** [comment]
> are you currently working on the port trapping feature?

No, to be honest. And there's no time schedule.

> have a requirement that to allow non-DPMI client (real mode program or something like that) to install a PM port trap :D

HDPMI is a bit picky - or elitist. It does not talk to real-mode programs. :)

**Comment by crazii** [comment]
> > are you currently working on the port trapping feature?
> 
> No, to be honest. And there's no time schedule.
> 
> > have a requirement that to allow non-DPMI client (real mode program or something like that) to install a PM port trap :D
> 
> HDPMI is a bit picky - or elitist. It does not talk to real-mode programs. :)

Is it possible to do it like EMM386 in ring0? or it is bad?

EIDT: my scenario is a VCPI client USB OPL3 driver that handles EMM386 port trap (minimal dependency), so I'd like to add port trap to PM mode using the same piece of code. but it's OK DPMI cannot do it, because the driver has a DPMI build too.
I was thinking that if HDPMI can provide an EMM386 compatible interface, then other real time programs can handle PM port trapping without much modification.
